File: backend/app/schemas/draft.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, List
from datetime import datetime
from enum import Enum
import uuid


class DraftStatus(str, Enum):
    CREATED = "created"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    ABANDONED = "abandoned"


class DraftType(str, Enum):
    SNAKE = "snake"
    LINEAR = "linear"


class ScoringType(str, Enum):
    STANDARD = "standard"
    PPR = "ppr"
    HALF_PPR = "half_ppr"


class DraftTeamBase(BaseModel):
    team_index: int = Field(..., ge=0, description="0-based position in draft order")
    team_name: str = Field(..., min_length=1, max_length=100)
    is_user: bool = Field(default=False, description="True if human user, False if AI bot")


class DraftTeamCreate(DraftTeamBase):
    pass


class DraftTeam(DraftTeamBase):
    id: uuid.UUID
    draft_session_id: uuid.UUID
    current_roster: Dict[str, int] = Field(default_factory=dict, description="Current roster composition")
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class DraftPickBase(BaseModel):
    round_number: int = Field(..., ge=1)
    pick_number: int = Field(..., ge=1, description="Overall pick number")
    team_pick_number: int = Field(..., ge=1, description="Pick number within the team")


class DraftPickCreate(DraftPickBase):
    team_id: uuid.UUID
    player_id: Optional[uuid.UUID] = None


class DraftPick(DraftPickBase):
    id: uuid.UUID
    draft_session_id: uuid.UUID
    team_id: uuid.UUID
    player_id: Optional[uuid.UUID] = None
    picked_at: Optional[datetime] = None
    pick_time_seconds: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class DraftSessionBase(BaseModel):
    num_teams: int = Field(..., ge=4, le=32, description="Number of teams in the draft")
    draft_type: DraftType
    scoring_type: ScoringType
    roster_positions: Dict[str, int] = Field(..., description="Roster position requirements")


class DraftSessionCreate(DraftSessionBase):
    teams: List[DraftTeamCreate] = Field(..., description="Teams participating in the draft")
    
    class Config:
        json_schema_extra = {
            "example": {
                "num_teams": 12,
                "draft_type": "snake",
                "scoring_type": "ppr",
                "roster_positions": {
                    "qb": 1,
                    "rb": 2,
                    "wr": 2,
                    "te": 1,
                    "flex": 1,
                    "k": 1,
                    "dst": 1,
                    "bench": 6
                },
                "teams": [
                    {"team_index": 0, "team_name": "User Team", "is_user": True},
                    {"team_index": 1, "team_name": "AI Team 1", "is_user": False}
                ]
            }
        }


class DraftSession(DraftSessionBase):
    id: uuid.UUID
    status: DraftStatus
    current_round: int = Field(default=1, ge=1)
    current_pick: int = Field(default=1, ge=1)
    current_team_index: int = Field(default=0, ge=0)
    total_rounds: int = Field(..., ge=1)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    
    # Optional relationships
    teams: Optional[List[DraftTeam]] = None
    picks: Optional[List[DraftPick]] = None
    
    class Config:
        from_attributes = True


class DraftSessionSummary(BaseModel):
    """Lightweight version of DraftSession for listing purposes"""
    id: uuid.UUID
    num_teams: int
    draft_type: DraftType
    scoring_type: ScoringType
    status: DraftStatus
    current_round: int
    current_pick: int
    total_rounds: int
    created_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class MakePickRequest(BaseModel):
    player_id: uuid.UUID
    
    class Config:
        json_schema_extra = {
            "example": {
                "player_id": "123e4567-e89b-12d3-a456-426614174000"
            }
        }


class DraftStateResponse(BaseModel):
    """Response containing current draft state"""
    draft_session: DraftSession
    current_team: DraftTeam
    available_players: List[Dict] = Field(default_factory=list, description="Players available to draft")
    recent_picks: List[DraftPick] = Field(default_factory=list, description="Recent draft picks")
    
    class Config:
        json_schema_extra = {
            "example": {
                "draft_session": {
                    "id": "123e4567-e89b-12d3-a456-426614174000",
                    "status": "in_progress",
                    "current_round": 2,
                    "current_pick": 13,
                    "current_team_index": 0
                },
                "current_team": {
                    "team_name": "User Team",
                    "is_user": True
                },
                "available_players": [],
                "recent_picks": []
            }
        }

File: frontend/src/types/draft.ts
export enum DraftStatus {
  CREATED = 'created',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  ABANDONED = 'abandoned'
}

export enum DraftType {
  SNAKE = 'snake',
  LINEAR = 'linear'
}

export enum ScoringType {
  STANDARD = 'standard',
  PPR = 'ppr',
  HALF_PPR = 'half_ppr'
}

export interface DraftTeam {
  id: string;
  draft_session_id: string;
  team_index: number;
  team_name: string;
  is_user: boolean;
  current_roster: Record<string, number>;
  created_at: string;
  updated_at: string;
}

export interface DraftPick {
  id: string;
  draft_session_id: string;
  team_id: string;
  player_id?: string;
  round_number: number;
  pick_number: number;
  team_pick_number: number;
  picked_at?: string;
  pick_time_seconds?: number;
  created_at: string;
  updated_at: string;
  // Additional fields from API
  player?: Player;
  team?: DraftTeam;
}

export interface DraftSession {
  id: string;
  num_teams: number;
  draft_type: DraftType;
  scoring_type: ScoringType;
  roster_positions: Record<string, number>;
  status: DraftStatus;
  current_round: number;
  current_pick: number;
  current_team_index: number;
  total_rounds: number;
  started_at?: string;
  completed_at?: string;
  created_at: string;
  updated_at: string;
  teams?: DraftTeam[];
  picks?: DraftPick[];
}

export interface Player {
  id: string;
  player_name: string;
  position: string;
  team: string;
  ecr_rank_ppr?: number;
  ecr_rank_standard?: number;
  ecr_rank_half_ppr?: number;
  adp_ppr?: number;
  adp_standard?: number;
  adp_half_ppr?: number;
  previous_year_points_ppr?: number;
  previous_year_points_standard?: number;
  previous_year_points_half_ppr?: number;
}

export interface DraftStateResponse {
  draft_session: DraftSession;
  current_team: DraftTeam;
  available_players: Player[];
  recent_picks: DraftPick[];
}

export interface MakePickRequest {
  player_id: string;
  player_name: string;
}

export interface DraftBoardProps {
  draftId: string;
}

export interface DraftTeamCreate {
  team_index: number;
  team_name: string;
  is_user: boolean;
}

export interface DraftSessionCreate {
  num_teams: number;
  draft_type: DraftType;
  scoring_type: ScoringType;
  roster_positions: Record<string, number>;
  teams: DraftTeamCreate[];
}

// UI-specific types
export interface DraftBoardCell {
  round: number;
  team_index: number;
  pick_number: number;
  pick?: DraftPick;
  is_current: boolean;
  is_user_team: boolean;
}

export interface PlayerSearchFilters {
  position?: string;
  search_text?: string;
  min_rank?: number;
  max_rank?: number;
}

export interface DraftOrderData {
  draft_order: number[][];
  team_names: string[];
  current_round: number;
  current_pick: number;
  current_team_index: number;
  total_rounds: number;
  draft_type: string;
  is_complete: boolean;
}